neqo-transport/src/fc.rs

```diff
@@ -904,4 +1017,215 @@ mod test {
     fn local_stream_limits_new_stream_server() {
         local_stream_limits(Role::Server, 1, 3);
     }
+
+    fn write_frames(fc: &mut ReceiverFlowControl<StreamId>, rtt: Duration, now: Instant) -> usize {
+        let mut builder = PacketBuilder::short(Encoder::new(), false, None::<&[u8]>);
+        let mut tokens = Vec::new();
+        fc.write_frames(
+            &mut builder,
+            &mut tokens,
+            &mut FrameStats::default(),
+            now,
+            rtt,
+        );
+        tokens.len()
+    }
+
+    #[test]
+    fn trigger_factor() -> Res<()> {
+        let rtt = Duration::from_millis(40);
+        let now = Instant::now();
+        let mut fc = ReceiverFlowControl::new(StreamId::new(0), INITIAL_RECV_WINDOW_SIZE as u64);
+
+        let fraction = INITIAL_RECV_WINDOW_SIZE as u64 / WINDOW_UPDATE_FRACTION;
+
+        let consumed = fc.set_consumed(fraction)?;
+        fc.add_retired(consumed);
+        assert_eq!(write_frames(&mut fc, rtt, now), 0);
+
+        let consumed = fc.set_consumed(fraction + 1)?;
+        assert_eq!(write_frames(&mut fc, rtt, now), 0);
+
+        fc.add_retired(consumed);
+        assert_eq!(write_frames(&mut fc, rtt, now), 1);
+
+        Ok(())
+    }
+
+    #[test]
+    fn auto_tuning_increase_no_decrease() -> Res<()> {
+        let rtt = Duration::from_millis(40);
+        let mut now = Instant::now();
+        let mut fc = ReceiverFlowControl::new(StreamId::new(0), INITIAL_RECV_WINDOW_SIZE as u64);
+        let initial_max_active = fc.max_active();
+
+        // Consume and retire multiple receive windows without increasing time.
+        for _ in 1..11 {
+            let consumed = fc.set_consumed(fc.next_limit())?;
+            fc.add_retired(consumed);
+            write_frames(&mut fc, rtt, now);
+        }
+        let increased_max_active = fc.max_active();
+
+        assert!(
+            initial_max_active < increased_max_active,
+            "expect receive window auto-tuning to increase max_active on 0 latency connection"
```

Comment:
Is this right?  I guess that the expectation is that the point of the test is that time has not advanced, therefore the RTT doesn't matter.  A comment about that seems prudent, because this is NOT a 0 latency connection, it's just that the elapsed time is zero.

Reply:
Good catch. Addressed in https://github.com/mozilla/neqo/pull/1868/commits/bbf09a561ef93be0c25ba0bc81008e1c92a86394.